allow empty responses

lag-of-death opened the pull request:
Fixes #584
Fixes #592

## Checklist

- [x] Tests have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## What kind of change does this PR introduce?

Bug fix.

## What is the current behavior? What is the new behavior?

No empty responses were allowed.

StefanDywersant commented on packages/http/src/mocker/negotiator/NegotiatorHelpers.ts:
```diff
@@ -142,12 +170,7 @@ const helpers = {
-                ProblemJsonError.fromTemplate(NOT_ACCEPTABLE, `Unable to find content for ${mediaTypes}`),
-              );
-            },
+            () => whenNoContent(logger, response, headers || [], mediaTypes),
```
You should also apply the support for empty content here: https://github.com/stoplightio/prism/blob/master/packages/http/src/mocker/negotiator/NegotiatorHelpers.ts#L316

It will let you "hit two birds with one stone". This [PR](https://github.com/stoplightio/prism/pull/609) introduces empty response for invalid requests, however you solution looks more generic.

If you feel that this is not in scope I'll create new PR based on yours.

lag-of-death replied:
thanks @StefanDywersant for pointing this out, will take care of this in this PR

XVincentX commented on packages/http/src/mocker/negotiator/NegotiatorHelpers.ts:
```diff
@@ -22,6 +23,30 @@ import {
+  logger: Logger,
+  response: IHttpOperationResponse,
+  headers: IHttpHeaderParam[],
+  mediaTypes: string[],
```
I would really specify the return type explicitly — it helps the readability but more importantly the type checker will help us in case we're going to refactor this function, effectively disabling the inference.

XVincentX commented on packages/http/src/mocker/negotiator/NegotiatorHelpers.ts:
```diff
@@ -22,6 +23,30 @@ import {
 } from './InternalHelpers';
 import { IHttpNegotiationResult, NegotiatePartialOptions, NegotiationOptions } from './types';
 
+export const warnMsg = (contentTypes: string[]) => `Unable to find content for ${contentTypes}`;
```
In general if we're not using this at least 2-3 times, let's refrain from creating another function.

lag-of-death replied:
1st use: https://github.com/stoplightio/prism/pull/606/files#diff-47034315930357c592d473ec0fe19afcR179
  
2nd use:  https://github.com/stoplightio/prism/pull/606/files#diff-47034315930357c592d473ec0fe19afcR184

XVincentX commented on packages/http/src/mocker/negotiator/NegotiatorHelpers.ts:
```diff
@@ -22,6 +23,30 @@ import {
+  if (isContentEmpty && wasAcceptHeaderLeftUnspecified) {
+    logger.warn(`${warnMsg(mediaTypes)}. Sending an empty response.`);
+
+    return Option.some({
```
I am assuming this is an `IHttpResponse` object?

XVincentX commented on packages/http/src/mocker/negotiator/NegotiatorHelpers.ts:
```diff
@@ -143,9 +168,18 @@ const helpers = {
-              return Either.left<Error, IHttpNegotiationResult>(
-                ProblemJsonError.fromTemplate(NOT_ACCEPTABLE, `Unable to find content for ${mediaTypes}`),
+              return pipe(
+                whenNoContent(logger, response, headers || [], mediaTypes),
```
The readability of this is not clear. `whenNoContent` — what happens?

XVincentX commented on packages/http/src/mocker/negotiator/NegotiatorHelpers.ts:
```diff
@@ -143,9 +168,18 @@ const helpers = {
+                      ProblemJsonError.fromTemplate(NOT_ACCEPTABLE, `Unable to find content for ${mediaTypes}`),
+                    );
+                  },
+                  r => Either.right(r),
```
You do not need to use `fold` if all you need is convert this to an `Either`

`Either.fromOption` is the function you're looking for.

XVincentX commented on packages/http/src/mocker/negotiator/NegotiatorHelpers.ts:
```diff
@@ -313,8 +347,25 @@ const helpers = {
+                          new Error(`Neither schema nor example defined for ${response.code} response.`),
+                        );
+                      },
+                      r => Either.right(r),
```
ditto, `Either.fromOption` is the function you should be using here.

lag-of-death commented on packages/http/src/mocker/HttpMocker.ts:
```diff
@@ -119,12 +119,13 @@ function assembleResponse(
         const mockedBody = computeBody(negotiationResult, payloadGenerator);
         const mockedHeaders = computeMockedHeaders(negotiationResult.headers || [], payloadGenerator);
 
+        const headers = negotiationResult.mediaType
```
This is to not have `Content-Type` when a response has no payload.

lag-of-death commented:
@XVincentX, I believe the feedback is addressed now, ready for another round

XVincentX commented:
@lag-of-death Can you rebase when you get a chance?

XVincentX commented on packages/http/src/mocker/negotiator/NegotiatorHelpers.ts:
```diff
@@ -153,9 +183,21 @@ const helpers = {
-              return Either.left<Error, IHttpNegotiationResult>(
-                ProblemJsonError.fromTemplate(NOT_ACCEPTABLE, `Unable to find content for ${mediaTypes}`),
+              return pipe(
+                optionallyGetPayloadlessResponse(response, headers || [], mediaTypes),
```
We should come up with a better function name, this one is kind of awkward.

lag-of-death replied:
@XVincentX, I'm all ears :)

lag-of-death commented on packages/http/src/mocker/negotiator/NegotiatorHelpers.ts:
```diff
@@ -23,6 +25,34 @@ import {
+      return Option.fromPredicate((c: { length: number }) => {
+        const isContentEmpty = c.length === 0;
+
+        return isContentEmpty && wasAcceptHeaderLeftUnspecified;
```
Sure :)
so our `c` is `contents.filter(c => !c.mediaType.includes('*/*')` and our `wasAcceptHeaderLeftUnspecified` is `mediaTypes.filter(mt => !mt.includes('*/*')).length === 0`, so we know that we are excluding `*/*` from the equation here.
So `isContentEmpty && wasAcceptHeaderLeftUnspecified` is saying:
`if there is no specific content type defined in the operation` AND `Accept header sent from the user was different than */* or left empty`, then later 
```
return {
        code: response.code,
        headers,
      };
```
which basically means `return a payloadless response`
Hope it clarifies ;)

lag-of-death commented on packages/http-server/src/__tests__/server.oas.spec.ts:
```diff
@@ -138,8 +138,7 @@ describe.each([['petstore.no-auth.oas2.yaml', 'petstore.no-auth.oas3.yaml']])('s
 
-    expect(response.statusCode).toBe(422);
-    checkErrorPayloadShape(response.payload);
+    expect(response.statusCode).toBe(400);
```
because now empty responses are allowed

XVincentX replied:
Oh whoa, this is really cool!

lag-of-death commented on packages/http/src/mocker/negotiator/__tests__/NegotiatorHelpers.spec.ts:
```diff
@@ -146,9 +175,7 @@ describe('NegotiatorHelpers', () => {
-          assertLeft(negotiationResult, e =>
-            expect(e.message).toBe('Neither schema nor example defined for 422 response.'),
-          );
+          assertRight(negotiationResult, e => expect(e).toStrictEqual({ code: '422', headers: [] }));
```
because it now can be an empty response

XVincentX commented on packages/http/src/mocker/negotiator/InternalHelpers.ts:
```diff
@@ -19,7 +19,9 @@ export function findExampleByKey(httpContent: IHttpContent, exampleKey: string)
-  return !!v.contents;
+  const x = v.contents || [];
+
+  return !!x.filter(y => !y.mediaType.includes('*/*')).length;
```
We're going to remove this hack when http-spec will be fixed, right?

lag-of-death replied:
Yes :)

XVincentX commented on packages/http/src/mocker/negotiator/NegotiatorHelpers.ts:
```diff
@@ -1,12 +1,14 @@
+import { IHttpHeaderParam } from '@stoplight/types';
 import * as Either from 'fp-ts/lib/Either';
 import { NonEmptyArray } from 'fp-ts/lib/NonEmptyArray';
+import { fromNullable } from 'fp-ts/lib/Option';
```
Ping

XVincentX commented on packages/http/src/mocker/negotiator/NegotiatorHelpers.ts:
```diff
@@ -170,6 +222,27 @@ const helpers = {
+                //     },
+                //     yo,
+                //   );
+                // }),
```
Can we remove this?

XVincentX commented on packages/http/src/mocker/negotiator/__tests__/NegotiatorHelpers.spec.ts:
```diff
@@ -527,13 +554,27 @@ describe('NegotiatorHelpers', () => {
       });
 
-      describe('httpContent does not exist ', () => {
+      describe('httpContent does not exist', () => {
```
Maybe we can specify that "the response exists, but there is no httpcontent"

XVincentX commented on test-harness/specs/E2E-Accept-Content-Negotiation-AC-7.oas2.txt:
```diff
@@ -7,6 +7,8 @@ swagger: "2.0"
   /todos:
     get:
+      produces:
+        - text/plain
```
Yes!

lag-of-death commented on packages/http/src/mocker/negotiator/NegotiatorHelpers.ts:
```diff
@@ -153,9 +183,21 @@ const helpers = {
+                Either.fromOption(() => {
+                  logger.warn(warnMsg(mediaTypes));
+
+                  return ProblemJsonError.fromTemplate(
```
hm.. weren't we doing this:
```typescript
return Either.left<Error, IHttpNegotiationResult>(
                ProblemJsonError.fromTemplate(NOT_ACCEPTABLE, `Unable to find content for ${mediaTypes}`),
```
wanted to stay consistent

XVincentX commented on packages/http/src/mocker/negotiator/NegotiatorHelpers.ts:
```diff
@@ -23,6 +24,42 @@ import {
+        const noAcceptAndNoContentResponses =
+          !noContentResponses.length && !mediaTypes.filter((mt: string) => !mt.includes('*/*')).length;
+
+        return noAcceptAndNoContentResponses || ['400', '422'].includes(response.code);
```
Why are you checking for 400 and 422 responses manually? Don't they follow the same flow?

XVincentX commented on packages/http/src/mocker/negotiator/NegotiatorHelpers.ts:
```diff
@@ -23,6 +24,42 @@ import {
+    Option.chain(responses => {
+      return Option.fromPredicate((noContentResponses: { length: number }) => {
+        const noAcceptAndNoContentResponses =
+          !noContentResponses.length && !mediaTypes.filter((mt: string) => !mt.includes('*/*')).length;
```
I really think we should put a comment on this condition, I still struggle to understand it correctly.

Also, with includes `Accept: application/json,*/*` will match. Is this intentional? I guess so (if the client accepts */* then an empty response is viable, right?)

XVincentX commented on packages/http/src/mocker/negotiator/__tests__/NegotiatorHelpers.spec.ts:
```diff
@@ -84,6 +95,24 @@ describe('NegotiatorHelpers', () => {
       });
 
       describe('and has no static contents', () => {
+        it('returns an empty payload response', () => {
```
Let's put a word or two about the fact we're looking for a response for an invalid Request

XVincentX commented on packages/http/src/mocker/negotiator/NegotiatorHelpers.ts:
```diff
@@ -35,14 +35,12 @@ const findEmptyResponse = (
-      })(responses);
-    }),
+    Option.chain(responses =>
+      Option.fromPredicate(
```
`pipe(responses, Option.fromPredicate())`

lag-of-death commented:
hi @XVincentX, I think I've addressed the feedback. The logic for `findEmptyResponse` is now greatly simplified. 
The previous version was kind of redundant because some of the conditional checks were already performed up the chain - somewhere in `findEmptyResponse`'s caller or higher.
There are also a couple of more tests so that hopefully no edge case is missed.

XVincentX commented on packages/http/src/mocker/negotiator/NegotiatorHelpers.ts:
```diff
@@ -23,6 +24,28 @@ import {
+  mediaTypes: string[],
+): Option.Option<IHttpNegotiationResult> => {
+  return pipe(
+    Option.fromPredicate((contentTypes: string[]) => {
```
`pipe(mediaTypes, Option.fromPredicate(…), Option.map`
